### ISSUE TITLE
[RabbitMQ] Fowarding from EnMasse to RabbitMQ fails with inequivalent arg 'durable' for queue

### ISSUE THREAD

**Issue by k-wall** [EnMasseProject/enmasse#4532]
**Describe the bug**

On trying to forward messages from EnMasse to RabbitMQ where the queue already exists on the RabbitMQ broker.  The AMQP session was ended in the following way.
 
```
2020-05-13 23:30:34.973 [warning] <0.920.0> Closing session for connection <0.665.0>:
{'v1_0.error',{symbol,<<"amqp:precondition-failed">>},{utf8,<<"PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'stocks' in vhost '/': received 'false' but current is 'true'">>},undefined}
```

```
2020-05-14 00:16:28.672300 +0000 PROTOCOL (trace) [3285]:FRAME: 0 <- @open(16) [container-id="rabbit@rhfed-localdomain", max-frame-size=16384, channel-max=32767, idle-time-out=60000, properties={:"cluster_name"="rabbit@rhfed-localdomain", :copyright="Copyright (c) 2007-2020 Pivotal Software, Inc.", :information="Licensed under the MPL 1.1. Website: https://rabbitmq.com", :platform="Erlang/OTP 22.3.2", :product="RabbitMQ", :version="3.8.3"}]
2020-05-14 00:16:28.672583 +0000 ROUTER (info) [C3285] Connection Opened: dir=out host=192.168.1.111:5671 vhost= encrypted=TLSv1.2 auth=EXTERNAL user=(null) container_id=rabbit@rhfed-localdomain props={:"cluster_name"="rabbit@rhfed-localdomain", :copyright="Copyright (c) 2007-2020 Pivotal Software, Inc.", :information="Licensed under the MPL 1.1. Website: https://rabbitmq.com", :platform="Erlang/OTP 22.3.2", :product="RabbitMQ", :version="3.8.3"}
2020-05-14 00:16:29.316774 +0000 PROTOCOL (trace) [3285]:FRAME: 0 -> @begin(17) [next-outgoing-id=0, incoming-window=2147483647, outgoing-window=2147483647]
2020-05-14 00:16:29.316886 +0000 PROTOCOL (trace) [3285]:FRAME: 0 -> @attach(18) [name="mymessagingspace.stocks.stocksforwarder.out@mymessagingspace.stocks", handle=0, role=false, snd-settle-mode=2, rcv-settle-mode=0, source=@source(40) [address="stocks", durable=2, expiry-policy=:"session-end", timeout=0, dynamic=false, outcomes=@PN_SYMBOL[:"amqp:accepted:list", :"amqp:rejected:list", :"amqp:released:list", :"amqp:modified:list"]], target=@target(41) [address="stocks", durable=0, expiry-policy=:"session-end", timeout=0, dynamic=false], initial-delivery-count=0, max-message-size=0]
2020-05-14 00:16:29.390459 +0000 PROTOCOL (trace) [3285]:FRAME: 0 <- @begin(17) [remote-channel=0, next-outgoing-id=0, incoming-window=65535, outgoing-window=65535, handle-max=4294967295]
2020-05-14 00:16:29.391276 +0000 PROTOCOL (trace) [3285]:FRAME: 0 <- @end(23) [error=@error(29) [condition=:"amqp:precondition-failed", description="PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'stocks' in vhost '/': received 'false' but current is 'true'"]]
2020-05-14 00:16:29.392760 +0000 PROTOCOL (trace) [3285]:FRAME: 0 -> @end(23) []
```

**To Reproduce**

**Expected behavior**

**Screenshots**

**Additional context**

**Comment by k-wall** [comment]
I think what is happening here is that RabbitMQ is erroneously treating the partner's desire for terminus durability 0 and a mandatory demand for a non-durable queue.  

This is contrary to 2.6.3 Establishing Or Resuming A Link":

When a link is established, it is possible for an endpoint not to have all the capabilities necessary to create the terminus exactly matching the expectations of the peer. If this happens, the endpoint MAY adjust the properties in order to succeed in creating the terminus. In this case the endpoint MUST report the actual properties of the terminus as created.

RabbitMQ ought to replying with an Attach response with the terminus durability 1  and not detaching the link

**Comment by k-wall** [comment]
I also notice when RabbitMQ closes the session, that EnMasse doesn't retry.  I think this is a robustness issue with the EnMasse Broker Plugin - it should be retrying the connection.

**Comment by rgodfrey** [comment]
@k-wall yeah terminus durability and queue durability are essentially orthogonal ... terminus durability is only defining what link state should be retained when the link is detached.  Thus this is clearly a RabbitMQ bug - you should always be able to create a link for which no state is retained on detach.  

As to retry - agree this is desirable

**Comment by k-wall** [comment]
@rgodfrey suggested that the plugin could potentially work around the issue by detecting Rabbit and retrying the link with different terminus durability setting.  I'll explore that.

**Comment by k-wall** [comment]
Working around this problem in the Broker Plugin is tricky. The Broker plugin has no way to detect that it is RabbitMQ is (ultimately) on the other end of the link.  Nor does it know that the router's attempt to form the link to RabbitMQ is failing with a `End{amqp:precondition-failed}`    

To workaround this problem, the only way I currently see is to add an additional configuration item to the forwarder spec:

```
   - direction: out
      name: stocksforwarder
      remoteAddress: rabbit/stocks
     terminusDurabilityRabbitMqWorkaround: 2
```

The user would need to set the `terminusDurabilityRabbitMqWorkaround` to match the durability of the queue on the RabbitMQ server. This is most unpleasant -  but I don't see an alternative.    @lulf @rgodfrey WDYT?

**Comment by lulf** [comment]
I'm ok with that fix, but I'm not sure we can/should rely on the amqp-connector plugin for 1.0 - I did remove the amqp-connector module with the intention that we do forwarding using builtin functionality of the router and broker (though I'm fine to bring it back if you think otherwise). 

I think there is ongoing work in the broker to support AMQP outgoing connectors. Alternatively we can rely on router autolinks for this purpose.... We should ensure that whatever mechanism we use, it would support the rabbitmq forwarding use case, but I was hoping we could get rid of our custom broker plugins.

**Comment by k-wall** [comment]
@lulf  For compatibility with the RabbitMQ plugin as it stands today, we will continue to need a mechanism to influence the terminusDurability of a link that's accessible from end-user configuration.   Either a Artemis plugin needs to implement this, or we could try to push the feature back to Artemis itself (after all, it will meet that same issue for federation with RabbitMQ use-cases).  I agree it is preferable not to keep the EnMasse Artemis Plugin.

Separately,there is a router defect that would need to be fixed to handle links to RabbitMQ robustly:  The RabbitMQ plugin `End` the session in a number of error cases.  Currently it seems that Router ignores session `End` so the end result is a non-functioning link route.   I think the router should be recreating the session and recreating the link within it.

**Comment by lulf** [comment]
@k-wall FWIW, if we were to try using router autolinks for this, it is possible to specify terminus durability of the autolink, so we would be able to do that I think. The downside with autolinks is of course the potential message ordering issues when using message routing.

**Comment by k-wall** [comment]
> @k-wall FWIW, if we were to try using router autolinks for this, it is possible to specify terminus durability of the autolink, so we would be able to do that I think. The downside with autolinks is of course the potential message ordering issues when using message routing.

True, I hadn't consider that. That would work for use-cases that don't require order.
